File: corelib/src/System/IO/InputStream.cpp
#include <QFile>
#include <QByteArray>

void System::IO::InputStream::OpenInt(int fd) {
  QFile *file = new QFile();
  if (!file->open(fd, QIODevice::ReadOnly | QIODevice::Unbuffered)) {
    throw new System::Exception();
  }
  Value = (void*)file;
}

void System::IO::InputStream::OpenString(System::String *filename) {
  QFile *file = new QFile(QString((QChar*)filename->Value->Array, filename->Value->Length));
  if (!file->open(QIODevice::ReadOnly)) {
    throw new System::Exception();
  }
  Value = (void*)file;
}

int System::IO::InputStream::ReadByteArray(Core::FixedArray$T<uint8> *array) {
  QFile *file = (QFile*)Value;
  return file->read((char*)array->Array, array->Length);
}

System::String* System::IO::InputStream::ReadString() {
  QFile *file = (QFile*)Value;
  QByteArray ba = file->readLine();
  return Core::utf8ToString((const char*)ba.constData());
}

int System::IO::InputStream::AvailableRead() {
  QFile *file = (QFile*)Value;
  return file->bytesAvailable();
}

File: corelib/src/System/IO/InputStream.cs
using System;

namespace System.IO {
  public class InputStream {
    public InputStream(int fd) {
      OpenInt(fd);
    }
    public InputStream(String filename) {
      OpenString(filename);
    }

    public int Read(byte[] array) {
      return ReadByteArray(array);
    }
    public String ReadLine() {
      return null;
    }
    public int Available() {
      return AvailableRead();
    }

    private unsafe void* Value;
    private extern void OpenInt(int fd);
    private extern void OpenString(String filename);
    private extern int ReadByteArray(byte[] array);
    private extern String ReadString();
    private extern int AvailableRead();
  }
}
